c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -454,10 +461,77 @@ struct ExpandableSegment {
     return rangeFromHandles(begin, end);
   }
 
+  // Setup IPC sharing for range.
+  // Returns the (larger) range that was actually shared.
+  // Serializes data to std::ostream that can be passed to the
+  // other process, and then restored as an exapandable segment
+  // via ExpandableSegment::fromShared(istream);
+  SegmentRange share(SegmentRange range, std::ostream& buf) {
+    auto begin = segmentLeft(range.ptr);
+    auto end = segmentRight(range.ptr + range.size);
+    ShareHeader header{getpid(), segment_size_, end - begin};
+    buf.write((const char*)&header, sizeof(ShareHeader));
+    for (auto i : c10::irange(begin, end)) {
+      auto& handle = handles_.at(i).value();
+      if (!handle.fd) {
+        int fd = 0;
+        C10_CUDA_DRIVER_CHECK(DriverAPI::get()->cuMemExportToShareableHandle_(
+            &fd, handle.handle, CU_MEM_HANDLE_TYPE_POSIX_FILE_DESCRIPTOR, 0));
+        handle.fd = fd;
+      }
+      int fd = *handle.fd;
+      buf.write((const char*)&fd, sizeof(int));
+    }
+    return rangeFromHandles(begin, end);
+  }
+
+  static std::unique_ptr<ExpandableSegment> fromShared(
+      c10::DeviceIndex device,
+      std::vector<c10::DeviceIndex> peers,
+      std::istream& buf) {
+    ShareHeader header{};
+    buf.read((char*)&header, sizeof(ShareHeader));
+    auto segment = std::make_unique<ExpandableSegment>(
+        device,
+        std::nullopt,
+        header.num_handles * header.segement_size,
+        header.segement_size,
+        std::move(peers));
+    auto pidfd = syscall(SYS_pidfd_open, header.pid, 0);
+    TORCH_CHECK(pidfd != -1, "pidfd_open:", std::strerror(errno));
+    for (auto i : c10::irange(header.num_handles)) {
+      (void)i;
+      int fd = 0;
+      buf.read((char*)&fd, sizeof(int));
+      auto myfd = syscall(SYS_pidfd_getfd, pidfd, fd, 0);
```

Comment:
Not well familiar with pidfd_getfd, but IIUC this create a duplicate fd (referring to same resource) from other process. Is that required? I thought the fd obtained via cuMemExportToShareableHandle can be directly plugged into cuMemImportFromShareableHandle (or are we worried about it being released in-between?)

Reply:
Since they are in different processes, they need to be different file descriptor numbers. The things that are needed are taken from the only same of this API: https://github.com/NVIDIA/cuda-samples/tree/master/Samples/3_CUDA_Features/memMapIPCDrv